src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -81,6 +81,92 @@ allocation_constraints partition_allocator::default_constraints() {
     return req;
 }
 
+bool guess_is_memory_group_aware_memory_limit() {
+    // Here we are trying to guess whether `topic_memory_per_partition` is
+    // memory group aware. Originally the property was more like a guess of how
+    // much memory each partition would statically use. It was first set to 10
+    // MiB and later reduced to 4MiB. Both of those weren't really factual and
+    // rather set such that they about align with `topic_partitions_per_shard`.
+    //
+    // These days we have a better understanding of how much memory a partition
+    // actually uses at rest based on memory profiler and metrics analysis.
+    // Hence we now set it to a more realistic value in the XXXKiB range.
+    //
+    // At the same time we also made the memory_groups reserve a percentage of
+    // the total memory space for partitions. Instead of dividing the whole
+    // memory space by `topic_memory_per_partition` we now only do that with the
+    // reserved value.
+    //
+    // Because of the latter making this change isn't trivial. Users might have
+    // lowered the `topic_memory_per_partition` property to gain higher
+    // partition density. With only using the reserved memory space when
+    // checking memory limits this might lead to now actually having less
+    // partition density. Hence, we need to guess whether the property was set
+    // before this change or after. If it was set before we just use the old
+    // rules (while still reserving a part of the memory) when checking memory
+    // limits to not break existing behaviour.
+    //
+    // Note that if we get this guess wrong that isn't immediately fatal. This
+    // only becomes active when creating new topics or modifying existing
+    // topics. At that point users can increase the memory reservation if
+    // needed.
+
+    // not overriden so definitely memory group aware
+    if (!config::shard_local_cfg().topic_memory_per_partition.is_overriden()) {
+        return true;
+    }
+
+    auto value = config::shard_local_cfg().topic_memory_per_partition.value();
+
+    // Previous default was 4MiB (and even higher before that). New one is more
+    // than 10x smaller. We assume it's unlikely someone would have changed the
+    // value to be that much smaller and hence guess that all the values larger
+    // than 2 times the new default are old values. Equally in the new world
+    // it's unlikely anybody would increase the value (at all really).
```

Comment:
I guess the most likely people are ourselves, if we find out that 100K or whatever is not enough, i.e., there is some other significant source of memory usage we didn't account for. If we need to go to 300K say, we can't because we'd be over this threshold and it would have the opposite as expected effect. However in this case we could just tweak the shares for a similar effect? E.g., reduce shares to 33% of their existing value (this is not really exactly the same since it also in theory gives that memory away to other subsystems but in practice based on how this works it seems similar).

Reply:
I mean we can always change the default up reallyas the heuristic is only really needed if the value is overriden (at which point we can also reevaluate the heuristic). 

Lets say we had to change it to 400K from the current 200K then today I feel like the heuristic as is would still be useful as I don't think anyone would have put it below 1MB (and the caveat that nothing immediately breaks still applies).